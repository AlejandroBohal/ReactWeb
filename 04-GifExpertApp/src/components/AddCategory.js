import React, { useState } from 'react'
import PropTypes from 'prop-types'

//Componente padre GifExpertApp
const AddCategory = ({setCategories}) => {

    //El inputValue vive en el state de la aplicación (recordar).

    const [inputValue, setInputValue] = useState('');
    // e.target.value será el nuevo valor del inputValue
    const handleInputChange = ({target:{value}}) => setInputValue(value);
    const handleSubmit = (e) =>{
        //previene el refresh
        e.preventDefault();

        if(inputValue.trim().length > 2){
            setCategories(categories => [inputValue,...categories] );
            setInputValue('');
        }else{
            alert('Minimun Category length 2 is required')
        }
    
    }
    //El fragment agrupa los elementos html y no es necesario si hay una etiqueta que cumpla esta función (form)
    return (
        <form onSubmit={handleSubmit}>
            <p>{inputValue}</p>
            <input  
                type="text" 
                placeholder="Type a category..."
                value={inputValue}
                onChange={ handleInputChange }
            />
        </form>
        
    )
}
AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired,
}

export default AddCategory

