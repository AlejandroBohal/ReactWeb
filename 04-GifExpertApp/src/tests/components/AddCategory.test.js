import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import React from 'react';
import AddCategory from '../../components/AddCategory';

describe('Testing AddCategory component', () => {
    // jest funciona como dom, por lo tanto esta funci칩n estar치 en el contexto del componente
    const setCategories = jest.fn(); 
    window.alert = () => {};

    let wrapper = shallow(<AddCategory setCategories={setCategories} />);
    beforeEach( () =>{
        jest.clearAllMocks();
        wrapper = shallow(<AddCategory setCategories={setCategories} />);    
    });

    test('should load <AddCategory /> correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });
    test('should change category imput', () => {
        const input = wrapper.find('input');
        const value = 'input test';
        const event = {target: { value:value,}}
        input.simulate('change',event);
        expect(wrapper.find('p').text().trim()).toBe(value);
    });
    test("shouldn't setCategories onSubmit", () => {
        const event = {preventDefault(){}};
        wrapper.find('form').simulate('submit',event);
        // como no hay inputvalue, la funci칩n no debi칩 llamarse.
        expect( setCategories ).not.toHaveBeenCalled();
    });
    test('should call setCategories and clear input', () => {
        const value = 'input test';
        const input = wrapper.find('input');
        const {value:inputValue} = input.props();
        input.simulate('change',{target: { value:value,}});
        wrapper.find('form').simulate('submit',{preventDefault(){}});
        expect( setCategories ).toHaveBeenCalled();
        expect( setCategories ).toHaveBeenCalledTimes(1);
        expect( setCategories ).toHaveBeenCalledWith(expect.any(Function));
        expect(inputValue).toBe('');        
    })
    
    
    
})
