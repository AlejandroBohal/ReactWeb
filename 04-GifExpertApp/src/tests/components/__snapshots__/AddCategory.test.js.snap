// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing AddCategory component should load <AddCategory /> correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    placeholder="Type a category..."
    type="text"
    value=""
  />
</form>
`;
